Add PageRank notebook (#111)
networkx/notebooks: content/algorithms/pagerank/pagerank.md

```{code-cell} ipython3
# imports
import networkx as nx
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
```

```{code-cell} ipython3
# run pagerank algorithm on a small graph
G = nx.karate_club_graph()
pagerank_dict = nx.pagerank(G)
```

```{code-cell} ipython3
# plot the graph such that node size is proportional to the pagerank value of node
nsize = [x*25000 for x in pagerank_dict.values()]
plt.figure(figsize=(10,10))
pos=nx.circular_layout(G)
nx.draw(G, pos, with_labels=True, node_size=nsize, node_color='yellow')
plt.show()
```

From the plot above, we observe that the node size (which is proportional to the pagerank of the node) is proportional to the number of edges connected to the node. 

+++

### Markov Chains

+++

Markov chains are mathematical models that represent systems that evolve over time in a probabilistic manner. These systems consist of a set of states, and transitions between these states occur randomly based on specific probabilities.

> The key characteristic of Markov chains is that the probability of transitioning from one state to another depends solely on the current state and is independent of the past states.

Author: > The key characteristic of Markov chains is that the probability of transitioning from one state to another depends solely on the current state and is independent of the past states.

Also, now that I re-read this sentence, I realize that there may also be confusion because here "state" refers to the state of the entire system of the Markov Chain. Whereas in the rest of the text "state" refers to the individual states that actually form the Markov chain. I say the following when introducing Markov chains - 

> These systems consist of a set of states, and transitions between these states occur randomly based on specific probabilities.

> The key characteristic of Markov chains is that the probability of transitioning from one state to another depends solely on the current state and is independent of the past states.

Reviewer: I think "state of being" is still confusing because it also uses the word "state". 

Perhaps something like:  `to another depends solely on the state probabilities at the current time and is not directly dependent on any previous state probabilities."

> The key characteristic of Markov chains is that the probability of transitioning from one state to another depends solely on the current state and is independent of the past states.

Author: Yes, this is much better!

Markov chains can be used to model the web network, particularly in the context of the PageRank algorithm.

> In this context, individual web pages are represented as states, and going from one state to another involves transition probabilities. This transition probability is determined by the number and quality of incoming links to that page. Pages with more incoming links are considered to have a higher probability of being visited.

Author: You're so right. I, myself, had such a hard time understanding this distinction, and it's weird I did not include it in the first draft. Thank you! 

As for the incoming/ outgoing links, the thing is the transition probability _to a page_ depends on the number and quality of incoming links to it. But when plotting the Markov chain, it is easier to explain in terms of outgoing links because each outgoing link from a page has equal probability of being visited. Maybe I should explain this in the notebook in a better way.

outdeg_dict = dict(H.out_degree)
transition_probabilities = {}
for (x, y) in H.edges:
    transition_probabilities[(x, y)] = round(1/outdeg_dict[x], 2)
edge_labels = my_draw_networkx_edge_labels(H, pos, edge_labels=transition_probabilities, rotate=False, rad = 0.15)
```

In the above plot, we model a simple Markov chain with 4 states. The transition probability is simply based on the outgoing links assuming that user is equally likely to transition to all those states i.e. if a state has 3 outgoing links, the transition probability of each of those will be 0.33.

Transition probability of going from state 1 to state 2 can be represented as $ P(1, 2) = 0.5 $.

```{code-cell} ipython3
P = np.zeros((4, 4))
P = np.array([[transition_probabilities.get((i, j), 0) for j in range(4)] for i in range(4)])
df = pd.DataFrame(P, columns=[i for i in range (4)], index=[i for i in range (4)])
df
```

To represent the transition probabilities, a transition matrix is constructed as shown above. It is a square matrix with its rows and columns corresponding to the states (web pages). Entry $ P_{i j} $ in matrix represents the transition probability of going from state $i$ to state $j$.

> The fundamental components of a Markov chain are the set of states, transition probabilities, and an initial state distribution. 

Author: I believe the long-run probability of being in each state is the output of computing the stationary distribution of a Markov chain and not inherently a component of it. Plus, there exist periodic and reducible Markov chains that will not have a stable long run probability distribution.

> The fundamental components of a Markov chain are the set of states, transition probabilities, and an initial state distribution. 

Reviewer: Yes -- I was using the word "component" as you do in the text. Not in the sense of a graph component, but in the sense of a part of the Markov chain process. Maybe the wording "The fundamental parts of a Markov chain...". Or with the wording I used, "The fundamental inputs to a Markov chain..."

This brute-force method of computing stationary distributions by performing the operation $\pi_{n+1} = \pi_{n}.P$ at every step is called the Power Iteration method. We continue the iterations till the difference between $\pi_{n+1}$ and $\pi_{n}$ becomes less than the tolerance threshold value.

There exist algebraic methods that can compute stationary distributions in a more direct manner. But practical experiments on large web networks have found that the brute-force method has the best performance  because most web networks have relatively fast convergence times (~50 iterations to reach a reasonable tolerance threshold).

+++

## Existence of Stationary Distributions

+++

<i>Do all Markov chains converge to a Stationary distribution?</i>

To answer this question, let us look at two examples.

```{code-cell}
I = nx.MultiDiGraph()
I.add_edges_from([(1, 1), (0, 0)])

plt.figure(figsize=(4, 2))

pos = {0: (1, 1), 1: (2, 1)}
nx.draw(
    I,
    pos,
    with_labels=True,
    node_size=1000,
    node_color="yellow",
    connectionstyle="arc3, rad = 0.1",
)
el = nx.draw_networkx_edge_labels(I, pos, edge_labels={(1, 1): 1.0, (0, 0): 1.0})
plt.title("G1", y=-0.2)
plt.show()
```

> Both states in above Markov chain have only one incoming link that too from itself. Thus, it contains isolated subsets of states that cannot be reached from the rest of the chain. This is called a <b><i>reducible</i></b> Markov chain 
> 

Reviewer: ```suggestion
Both states in above Markov chain have only one incoming link that comes from itself. Thus, it contains isolated subsets of states that cannot be reached from the rest of the chain. This is called a <b><i>reducible</i></b> Markov chain.

```

> Here the transition probability matrix is an identity matrix because of which every initial distribution becomes a stationary distribution.

Reviewer: ```suggestion
Here the transition probability matrix is an identity matrix because every initial distribution is a stationary distribution. The state are sometimes called "absorbing states" because the system never leaves those states.
```

```{code-cell}
J = nx.DiGraph()
J.add_edges_from([(0, 1), (1, 0)])

plt.figure(figsize=(4, 2))

pos = {0: (1, 1), 1: (2, 1)}
nx.draw(
    J,
    pos,
    with_labels=True,
    node_size=1000,
    node_color="yellow",
    connectionstyle="arc3, rad = 0.2",
)
edge_labels = my_draw_networkx_edge_labels(
    J, pos, edge_labels={(0, 1): 1.0, (1, 0): 1.0}, rotate=False, rad=0.2
)

plt.title("G2", y=-0.1)
plt.show()
```

In the above Markov chain, when we apply transitions, the distribution never converges and only oscillates. This is called a <b><i>periodic</i></b> Markov chain.

Periodic Markov chains are characterized by having subsets of states that form closed loops or cycle, where transitions within the cycle are more probable than transitions between cycles. These loops introduce a regularity in the chain's behavior, causing the system to exhibit repetitive patterns as it evolves over time.

> A periodic Markoc chain does not converge to a stationary distribution.

Reviewer: ```suggestion
A periodic Markov chain does not converge to a stationary distribution. It oscillates between distributions.
```

<b>Reducibility (Uniqueness problem):</b> For a unique stationary distribution $\pi$ to exist, the Markov chain must not be reducible. The Markov chain must be irreducible i.e it is possible to reach any state from any other state, either in a single step or over multiple steps.

<b>Periodicity (Convergence problem):</b> For all initial distributions $\pi_{0}$ to converge to $\pi$, the Markov chain must not be periodic. The Markov chain must be aperiodic i.e. when we apply transitions, the distribution across states should not oscillate but converge after sufficient time.

When a Markov chain is both irreducible and aperiodic, it is guaranteed to have a unique stationary distribution. This distribution represents the long-term behavior of the system, where the probabilities of being in each state stabilize and remain constant over time.

+++

## PageRank Algorithm

The Pagerank algorithm is fundamentally based on the principle of computing Stationary Distributions of Markov chains. Our methods for computing stationary distributions rely on the assumption that the web network is a reducible and aperiodic Markov chain. However it may not always be true in reality.

> There maybe webpages that have no outgoing links, so all users will be absorbed into this state after some transitions (these are called sink states). There maybe two or more webpages linked to each other such that once a user enters, they get stuck in a never ending cycle and all users eventually end up cycling between these two states.
> 

Reviewer: ```suggestion
There may be webpages that have no outgoing links, so all users will be absorbed into this state after some transitions (these are called sink states). There maybe two or more webpages linked to each other such that once a user enters, they get stuck in a never ending cycle and all users eventually end up cycling between these two states. Other configurations of states can cause similarly odd behavior.

```

<i> So, how do we handle Markov chains that don't satisfy the required conditions?</i>

> At every step, we transition to a random state with the probability $\frac{1-\alpha}{N}$ where $N = $ Number of states and we can pick any value for $\alpha$.
> 
> But it is not sufficient to just add these random transition probabilities because it breaks the assumption of a probability distribution of across states since the distribution no longer sums to one. To solve this problem, we use the damping factor $\alpha$. We first reduce the original transition probability by $\alpha$ and then add the random state transition probability of $\frac{1-\alpha}{N}$.

Reviewer: ```suggestion
```suggestion
We change the random walk ever so slightly. At every step, take a random walk, as described above, with chance $\alpha$, and with chance $(1-\alpha)$ we "jump" to a uniformly chosen random state with the probability $\frac{1-\alpha}{N}$ for each of the $N$ states. Typically $\alpha$ is chosen to be about 0.85. The jumping ensures that any state can be reached from any other (irreducible) and that the jumps can happen at any timestep (aperiodic). Note that the sum of the probabilities is 1 as it should be.
```
```
